lighthouse-core/test/lib/traces/tracing-processor-test.js

```diff
@@ -205,20 +188,63 @@ describe('TracingProcessor lib', () => {
     });
   });
 
-  describe('risk to responsiveness', () => {
+  describe('getMainThreadTopLevelEvents', () => {
+    it('gets durations of top-level tasks', () => {
+      TracingProcessor = require('../../../lib/traces/tracing-processor');
+      const trace = {traceEvents: pwaTrace};
+      const tabTrace = new TraceOfTab().compute_(trace);
+      const ret = TracingProcessor.getMainThreadTopLevelEvents(tabTrace);
+
+      assert.equal(ret.length, 645);
+    });
+
+    it('filters events based on start and end times', () => {
+      TracingProcessor = require('../../../lib/traces/tracing-processor');
+      const baseTime = 20000 * 1000;
+      const name = 'TaskQueueManager::ProcessTaskFromWorkQueue';
+      const tabTrace = {
+        navigationStartEvt: {ts: baseTime},
+        mainThreadEvents: [
+          // 15ms to 25ms
+          {ts: baseTime + 15 * 1000, dur: 10 * 1000, name},
+          // 40ms to 60ms
+          {ts: baseTime + 40 * 1000, dur: 20 * 1000, name},
+          // 1000ms to 2000ms
+          {ts: baseTime + 1000 * 1000, dur: 1000 * 1000, name},
+          // 4000ms to 4020ms
+          {ts: baseTime + 4000 * 1000, dur: 20 * 1000, name},
+        ],
+      };
+
+      const ret = TracingProcessor.getMainThreadTopLevelEvents(
+        tabTrace,
+        50,
+        1500
+      );
+      assert.equal(ret.length, 2);
+      assert.equal(ret[0].start, 40);
+      assert.equal(ret[0].end, 60);
+      assert.equal(ret[0].duration, 20);
+      assert.equal(ret[1].start, 1000);
+      assert.equal(ret[1].end, 2000);
+      assert.equal(ret[1].duration, 1000);
+    });
+  });
+
+  describe('getMainThreadTopLevelEventDurations', () => {
     it('gets durations of top-level tasks', () => {
-      const tracingProcessor = new TracingProcessor();
-      const model = tracingProcessor.init(pwaTrace);
+      TracingProcessor = require('../../../lib/traces/tracing-processor');
       const trace = {traceEvents: pwaTrace};
-      const ret = TracingProcessor.getMainThreadTopLevelEventDurations(model, trace);
+      const tabTrace = new TraceOfTab().compute_(trace);
+      const ret = TracingProcessor.getMainThreadTopLevelEventDurations(tabTrace);
       const durations = ret.durations;
 
       function getDurationFromIndex(index) {
         return Number(durations[index].toFixed(2));
       }
 
       assert.equal(durations.filter(dur => isNaN(dur)).length, 0, 'NaN found');
-      assert.equal(durations.length, 652);
+      assert.equal(durations.length, 645);
```

Comment:
ah found the 7th it's a result of `!dur` they return an event that has no duration which we'll ignore